fendermint/vm/topdown/proof-service/src/service.rs

```diff
@@ -0,0 +1,281 @@
+
+        loop {
+            poll_interval.tick().await;
+
+            tracing::debug!("Poll interval tick");
+            if let Err(e) = self.generate_next_proofs().await {
+                tracing::error!(
+                    error = %e,
+                    "Failed to generate proofs, will retry on next tick"
+                );
+            }
+        }
+    }
+
+    /// Generate proofs for next needed instances
+    ///
+    /// Called by run() on each tick. Implements the core flow:
+    /// FETCH → VALIDATE → GENERATE → CACHE
+    ///
+    /// CRITICAL: Processes F3 instances SEQUENTIALLY - never skips!
+    async fn generate_next_proofs(&self) -> Result<()> {
+        let last_committed = self.cache.last_committed_instance();
+        // Lookahead window starts AFTER last_committed (which was already processed)
+        let next_instance = last_committed + 1;
+        let max_instance = last_committed + self.config.lookahead_instances;
+
+        tracing::debug!(
+            last_committed,
+            next_instance,
+            max_instance,
+            "Checking for new F3 certificates"
+        );
+
+        // Process instances IN ORDER - this is critical for F3
+        for instance_id in next_instance..=max_instance {
+            // Skip if already cached
+            if self.cache.contains(instance_id) {
+                tracing::debug!(instance_id, "Proof already cached");
+                continue;
+            }
+
+            // ====================
+            // STEP 1: FETCH + VALIDATE certificate (single operation!)
+            // ====================
+            let validated = match self.f3_client.fetch_and_validate(instance_id).await {
+                Ok(cert) => cert,
+                Err(e)
+                    if e.to_string().contains("not found")
+                        || e.to_string().contains("not available") =>
+                {
+                    // Certificate not available yet - STOP HERE!
+                    // Don't try higher instances as they depend on this one
+                    tracing::debug!(instance_id, "Certificate not available, stopping lookahead");
+                    break;
+                }
+                Err(e) => {
+                    return Err(e).with_context(|| {
+                        format!(
+                            "Failed to fetch and validate certificate for instance {}",
+                            instance_id
+                        )
+                    });
+                }
+            };
+
+            tracing::info!(
+                instance_id,
+                ec_chain_len = validated.f3_cert.ec_chain.suffix().len(),
+                "Certificate fetched and validated successfully"
+            );
+
+            // ====================
+            // STEP 2: GENERATE proof bundle
+            // ====================
+            let proof_bundle = self
+                .generate_proof_for_certificate(&validated.f3_cert)
```

Comment:
I'm not sure this needs to be executed separately for every instance/certificate.